### ISSUE TITLE
[Phase 7] Experiment Framework & CLI

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/multi-label-doc-classifier#8]
## Summary

Build the experiment framework and command-line interface for running reproducible experiments, training models, and making predictions.

**Parent Issue**: #1
**Depends On**: All previous phases

## Objectives

- Create YAML-based configuration system
- Build experiment runner with reproducibility
- Implement learning curve visualization
- Create CLI for train/predict/active-learn commands

## System Context

```
src/
├── cli.py                    # NEW - Command-line interface
└── experiments/
    ├── runner.py             # NEW - Experiment orchestration
    └── logging.py            # NEW - Result tracking

configs/
└── default.yaml              # NEW - Default configuration
```

### Configuration System
- [ ] Create `configs/default.yaml` with all parameters
- [ ] Implement config loading and validation
- [ ] Support config overrides from CLI
- [ ] Environment variable substitution

### Experiment Runner
- [ ] Create `src/experiments/runner.py`
- [ ] Implement `ExperimentRunner` class
- [ ] Handle random seed for reproducibility
- [ ] Save/load experiment checkpoints
- [ ] Generate learning curves

### Result Logging
- [ ] Create `src/experiments/logging.py`
- [ ] Implement `ExperimentLogger` class
- [ ] Save results to JSON/CSV
- [ ] Track hyperparameters and metrics
- [ ] Compare experiments

### CLI
- [ ] Create `src/cli.py` with argparse/click
- [ ] Implement `train` command
- [ ] Implement `predict` command
- [ ] Implement `active-learn` command
- [ ] Implement `evaluate` command

### Default Configuration

```yaml
# configs/default.yaml
experiment:
  name: "default_experiment"
  random_seed: 42
  output_dir: "results"

data:
  train_path: "data/raw/train.json"
  test_path: "data/raw/test.json"
  taxonomy_path: "data/labels/taxonomy.json"

preprocessing:
  lowercase: true
  remove_punctuation: true
  remove_numbers: false
  min_token_length: 2

vectorizer:
  max_features: 5000
  ngram_range: [1, 2]
  min_df: 2
  max_df: 0.95

classifier:
  type: "classifier_chain"  # or "label_powerset"
  base_estimator: "logistic_regression"
  base_estimator_params:
    max_iter: 1000
    C: 1.0

active_learning:
  enabled: true
  strategy: "uncertainty"  # or "committee", "hybrid"
  uncertainty_method: "entropy"
  initial_samples: 100
  samples_per_iteration: 20
  max_iterations: 50

semi_supervised:
  enabled: false
  confidence_threshold: 0.9
  max_iterations: 10

hierarchical:
  enabled: false
  enforce_constraints: true

evaluation:
  cross_validation_folds: 5
  metrics: ["hamming_loss", "subset_accuracy", "micro_f1", "macro_f1"]
```

### ExperimentRunner

```python
# src/experiments/runner.py
import yaml
import numpy as np
import json
from pathlib import Path
from datetime import datetime
from typing import Dict, Any, Optional

from src.features.preprocessor import TextPreprocessor
from src.features.vectorizer import DocumentVectorizer
from src.models.classifier_chain import ClassifierChain
from src.active_learning.uncertainty import UncertaintySampler
from src.active_learning.pool import UnlabeledPool
from src.evaluation.metrics import multilabel_report

class ExperimentRunner:
    """Orchestrates experiments with full reproducibility."""

    def __init__(self, config_path: str):
        """Load configuration and set up experiment.

        Args:
            config_path: Path to YAML config file
        """
        self.config = self._load_config(config_path)
        self.results = []

        # Set random seed
        seed = self.config['experiment']['random_seed']
        np.random.seed(seed)

        # Create output directory
        self.output_dir = Path(self.config['experiment']['output_dir'])
        self.output_dir.mkdir(parents=True, exist_ok=True)

        # Initialize components
        self._init_components()

    def _load_config(self, path: str) -> Dict:
        """Load and validate configuration."""
        with open(path, 'r') as f:
            config = yaml.safe_load(f)
        return config

    def _init_components(self):
        """Initialize all pipeline components."""
        prep_config = self.config['preprocessing']
        self.preprocessor = TextPreprocessor(
            lowercase=prep_config['lowercase'],
            remove_punctuation=prep_config['remove_punctuation'],
            remove_numbers=prep_config['remove_numbers'],
            min_token_length=prep_config['min_token_length']
        )

        vec_config = self.config['vectorizer']
        self.vectorizer = DocumentVectorizer(
            max_features=vec_config['max_features'],
            ngram_range=tuple(vec_config['ngram_range']),
            min_df=vec_config['min_df'],
            max_df=vec_config['max_df']
        )

    def run_training(self, X_train, y_train) -> Dict[str, Any]:
        """Run full training pipeline.

        Returns:
            Training results and metrics
        """
        # Preprocess and vectorize
        clean_texts = self.preprocessor.process_batch(X_train)
        X_vectorized = self.vectorizer.fit_transform(clean_texts)

        # Train classifier
        clf_config = self.config['classifier']
        if clf_config['type'] == 'classifier_chain':
            self.classifier = ClassifierChain()

        self.classifier.fit(X_vectorized, y_train)

        return {'status': 'trained', 'n_samples': len(X_train)}

    def run_active_learning(
        self,
        X_initial: np.ndarray,
        y_initial: np.ndarray,
        X_pool: np.ndarray,
        y_pool_hidden: np.ndarray,
        X_test: np.ndarray,
        y_test: np.ndarray
    ) -> Dict[str, Any]:
        """Run active learning experiment.

        Returns:
            Learning curves and final metrics
        """
        al_config = self.config['active_learning']

        # Preprocess all
        X_initial_clean = self.preprocessor.process_batch(X_initial)
        X_pool_clean = self.preprocessor.process_batch(X_pool)
        X_test_clean = self.preprocessor.process_batch(X_test)

        # Vectorize
        X_train_vec = self.vectorizer.fit_transform(X_initial_clean)
        X_pool_vec = self.vectorizer.transform(X_pool_clean)
        X_test_vec = self.vectorizer.transform(X_test_clean)

        # Setup active learning
        pool = UnlabeledPool(X_pool_vec, y_pool_hidden)
        sampler = UncertaintySampler(strategy=al_config['uncertainty_method'])

        # Initial training
        self.classifier = ClassifierChain()
        X_current = X_train_vec.copy()
        y_current = y_initial.copy()
        self.classifier.fit(X_current, y_current)

        learning_curve = []

        for iteration in range(al_config['max_iterations']):
            # Evaluate
            y_pred = self.classifier.predict(X_test_vec)
            metrics = multilabel_report(y_test, y_pred)
            metrics['iteration'] = iteration
            metrics['n_labeled'] = len(y_current)
            learning_curve.append(metrics)

            # Select samples
            if pool.n_unlabeled == 0:
                break

            X_unlabeled = pool.get_unlabeled()
            n_to_select = min(al_config['samples_per_iteration'], pool.n_unlabeled)
            selected_local = sampler.select_samples(
                self.classifier, X_unlabeled, n_to_select
            )
            selected_global = pool.unlabeled_indices[selected_local]

            # Get labels and update
            new_labels = pool.get_labels(selected_global)
            pool.mark_labeled(selected_global)

            # Add to training set
            from scipy.sparse import vstack
            X_current = vstack([X_current, X_pool_vec[selected_global]])
            y_current = np.vstack([y_current, new_labels])

            # Retrain
            self.classifier = ClassifierChain()
            self.classifier.fit(X_current, y_current)

        return {
            'learning_curve': learning_curve,
            'final_metrics': learning_curve[-1] if learning_curve else {}
        }

    def save_results(self, results: Dict[str, Any], name: str = 'results'):
        """Save experiment results to file."""
        timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
        output_path = self.output_dir / f"{name}_{timestamp}.json"

        with open(output_path, 'w') as f:
            json.dump(results, f, indent=2, default=str)

        print(f"Results saved to: {output_path}")
```

### CLI

```python
# src/cli.py
"""Command-line interface for multi-label document classifier."""

import argparse
import json
import sys
from pathlib import Path

def main():
    parser = argparse.ArgumentParser(
        description='Multi-Label Document Classifier with Active Learning'
    )
    subparsers = parser.add_subparsers(dest='command', help='Commands')

    # Train command
    train_parser = subparsers.add_parser('train', help='Train a classifier')
    train_parser.add_argument('--config', type=str, default='configs/default.yaml',
                             help='Path to config file')
    train_parser.add_argument('--data', type=str, required=True,
                             help='Path to training data')
    train_parser.add_argument('--output', type=str, default='models/model.pkl',
                             help='Path to save trained model')

    # Predict command
    predict_parser = subparsers.add_parser('predict', help='Make predictions')
    predict_parser.add_argument('--model', type=str, required=True,
                               help='Path to trained model')
    predict_parser.add_argument('--input', type=str, required=True,
                               help='Path to input documents')
    predict_parser.add_argument('--output', type=str, default='predictions.json',
                               help='Path to save predictions')

    # Active learning command
    al_parser = subparsers.add_parser('active-learn', help='Run active learning')
    al_parser.add_argument('--config', type=str, default='configs/default.yaml',
                          help='Path to config file')
    al_parser.add_argument('--labeled', type=str, required=True,
                          help='Path to initial labeled data')
    al_parser.add_argument('--unlabeled', type=str, required=True,
                          help='Path to unlabeled pool')
    al_parser.add_argument('--iterations', type=int, default=50,
                          help='Number of active learning iterations')
    al_parser.add_argument('--batch-size', type=int, default=20,
                          help='Samples to select per iteration')

    # Evaluate command
    eval_parser = subparsers.add_parser('evaluate', help='Evaluate model')
    eval_parser.add_argument('--model', type=str, required=True,
                            help='Path to trained model')
    eval_parser.add_argument('--test', type=str, required=True,
                            help='Path to test data')

    args = parser.parse_args()

    if args.command is None:
        parser.print_help()
        sys.exit(1)

    if args.command == 'train':
        from src.experiments.runner import ExperimentRunner
        from src.utils.data_loader import load_documents
        import joblib

        runner = ExperimentRunner(args.config)
        docs, labels = load_documents(args.data)
        results = runner.run_training(docs, labels)

        # Save model
        output_path = Path(args.output)
        output_path.parent.mkdir(parents=True, exist_ok=True)
        joblib.dump({
            'classifier': runner.classifier,
            'vectorizer': runner.vectorizer,
            'preprocessor': runner.preprocessor
        }, output_path)

        print(f"Model saved to: {output_path}")
        print(f"Results: {results}")

    elif args.command == 'predict':
        import joblib
        from src.utils.data_loader import load_documents

        # Load model
        model_data = joblib.load(args.model)
        classifier = model_data['classifier']
        vectorizer = model_data['vectorizer']
        preprocessor = model_data['preprocessor']

        # Load and process input
        docs = load_documents(args.input, labels=False)
        clean_docs = preprocessor.process_batch(docs)
        X = vectorizer.transform(clean_docs)

        # Predict
        predictions = classifier.predict(X)

        # Save predictions
        with open(args.output, 'w') as f:
            json.dump({'predictions': predictions.tolist()}, f, indent=2)

        print(f"Predictions saved to: {args.output}")

    elif args.command == 'active-learn':
        from src.experiments.runner import ExperimentRunner
        from src.utils.data_loader import load_documents

        runner = ExperimentRunner(args.config)

        # Override config with CLI args
        runner.config['active_learning']['max_iterations'] = args.iterations
        runner.config['active_learning']['samples_per_iteration'] = args.batch_size

        # Load data
        X_labeled, y_labeled = load_documents(args.labeled)
        X_unlabeled, y_unlabeled = load_documents(args.unlabeled)

        # Run active learning (using unlabeled as test for demo)
        results = runner.run_active_learning(
            X_labeled, y_labeled,
            X_unlabeled, y_unlabeled,
            X_unlabeled[:100], y_unlabeled[:100]  # Use subset as test
        )

        runner.save_results(results, 'active_learning')
        print("Active learning complete!")

    elif args.command == 'evaluate':
        import joblib
        from src.utils.data_loader import load_documents
        from src.evaluation.metrics import multilabel_report

        model_data = joblib.load(args.model)
        docs, labels = load_documents(args.test)

        clean_docs = model_data['preprocessor'].process_batch(docs)
        X = model_data['vectorizer'].transform(clean_docs)
        predictions = model_data['classifier'].predict(X)

        report = multilabel_report(labels, predictions)
        print("\nEvaluation Results:")
        for metric, value in report.items():
            print(f"  {metric}: {value:.4f}")

if __name__ == '__main__':
    main()
```

## Verification

```bash
# Test CLI commands
python -m src.cli train --config configs/default.yaml --data data/raw/train.json

python -m src.cli predict --model models/model.pkl --input data/raw/test.json

python -m src.cli active-learn --labeled data/raw/train.json \
    --unlabeled data/raw/unlabeled.json --iterations 10

python -m src.cli evaluate --model models/model.pkl --test data/raw/test.json

# Verify reproducibility
python -m src.cli train --config configs/default.yaml --data data/raw/train.json
# Run again with same config - should produce identical results
```

## Definition of Done

- [ ] Config loads and validates all parameters
- [ ] CLI commands work as documented
- [ ] Experiments are reproducible with same random seed
- [ ] Results are saved and can be loaded
- [ ] Learning curves can be visualized
- [ ] Error messages are helpful

## Files to Modify

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `src/cli.py` | - | Create | Command-line interface |
| `src/experiments/runner.py` | - | Create | Experiment orchestration |
| `src/experiments/logging.py` | - | Create | Result logging |
| `configs/default.yaml` | - | Create | Default configuration |
| `tests/test_cli.py` | - | Create | CLI integration tests |
| `tests/test_runner.py` | - | Create | Runner unit tests |
